File: utils/ADJOINT_TOMOGRAPHY_TOOLS/measure_adj/scripts_meas/plot_mtm.pl
#!/usr/bin/perl

use Time::Local;
use Getopt::Std;
use POSIX;
use lib '/opt/seismo-util/lib/perl';
use SACLST;
use File::Basename;
sub Usage{
  print STDERR <<END;

  plot_mtm.pl mtmfiles.dlnA
  EXAMPLE (execute from OUTPUT_FILES):
     ../scripts_meas/plot_mtm.pl *.dlnA

  original coding by Vala, adapted by Qinya for the MT_MEASURE_ADJ package

END
  exit(1);
}

if (@ARGV == 0) { Usage(); }
if (!getopts('i:')) {die('Check input arguments\n');}
#if($opt_i){$inputdir = $opt_i;}else{$inputdir = ".";}
#$tp = "mtm";

#@data_files=`ls -1 $inputdir/*.*.*.mtm.dlnA`;
@dlnA_files = @ARGV;

# USER MUST CHANGE THIS PATH
#$progdir = "/home/lqy/SVN/mt_measure_adj/scripts";
#$progdir = "/opt/seismo-util/source/multitaper";
$progdir = "/net/denali/scratch1/carltape/svn/cig/seismo/3D/ADJOINT_TOMO/measure_adj_work/scripts_meas";

my(@inputdirs);

#open(MATLAB,"> plot1.m");
# NOTE: Type "which matlab" to make sure that matlab is not alised.
open(MATLAB,"|/opt/matlab/bin/matlab -nosplash -nojvm") || die("Error opening matlab program\n");
#open(MATLAB,"|/usr/local/bin/matlab -nosplash -nojvm") || die("Error opening matlab program\n");

print MATLAB "P=path; path(P,'$progdir');\n";
print MATLAB "amps=[];phs=[];dlnAs=[];dts=[];gcarcs=[];azs=[];stats=[];\n";

print MATLAB "stats=char('first line');\n";

$iall = 1;  # 1: plot full transfer functions; 0: plot only average measurements

for ($i=0;$i< @dlnA_files;$i++) {

  ($filebase) = split("dlnA",`basename $dlnA_files[$i]`); chop($filebase);
  ($inputdir) = split(" ",`dirname $dlnA_files[$i]`);
  `sac2ascii.pl $inputdir/$filebase.obs.sac $inputdir/$filebase.syn.sac $inputdir/$filebase.recon_syn.sac $inputdir/$filebase.recon_syn_dt.sac $inputdir/$filebase.recon_syn_cc.sac $inputdir/$filebase.recon_syn_cc_dt.sac`;
  $epsfile = "$inputdir/${filebase}_measure.eps";

  (undef,$stname,$stnetwk,$comp,$az,$gcarc)=split(" ",`saclst kstnm knetwk kcmpnm az gcarc f $inputdir/${filebase}.obs.sac`);

  print MATLAB "disp(['$filebase'])\n";
  print MATLAB "data=dlmread('$inputdir/$filebase.obs');\n";
  print MATLAB "syn=dlmread('$inputdir/$filebase.syn');\n";
  print MATLAB "dataf=dlmread('$inputdir/$filebase.obs.power');\n";
  print MATLAB "synf=dlmread('$inputdir/$filebase.syn.power');\n";
  print MATLAB "new=dlmread('$inputdir/$filebase.recon_syn');\n";
  print MATLAB "new_dt=dlmread('$inputdir/$filebase.recon_syn_dt');\n";
  if ($iall == 1) {
    # read sub-sampled error curves
    print MATLAB "dlnA=dlmread('$inputdir/$filebase.err_dlnA');\n";
    print MATLAB "dt  =dlmread('$inputdir/$filebase.err_dt');\n";

    # read NON-sub-sampled error curves
    print MATLAB "dlnA_full=dlmread('$inputdir/$filebase.err_dlnA_full');\n";
    print MATLAB "dt_full  =dlmread('$inputdir/$filebase.err_dt_full');\n";

  } else {
    print MATLAB "dlnA=0; dlnA_full=0; dt=0; dt_full=0;\n";
  }

  # read frequency limits
  print MATLAB "flims=dlmread('$inputdir/$filebase.freq_limits');\n";

  # read multitaper average measurements
  print MATLAB "dlnA_ave=dlmread('$inputdir/$filebase.dlnA_average');\n";
  print MATLAB "dt_ave  =dlmread('$inputdir/$filebase.dt_average');\n";

  # read cross-correlation measurements
  print MATLAB "dlnA_cc=dlmread('$inputdir/$filebase.dlnA_cc');\n";
  print MATLAB "dt_cc  =dlmread('$inputdir/$filebase.dt_cc');\n";

  # read cross-correlation reconstruction
  print MATLAB "new_cc=dlmread('$inputdir/$filebase.recon_syn_cc');\n";
  print MATLAB "new_cc_dt=dlmread('$inputdir/$filebase.recon_syn_cc_dt');\n";

  print MATLAB "title = '$stname.$stnetwk.$comp \Delta = $gcarc^\\circ Az = $az^\\circ';\n";
  print MATLAB "figure;\n";

  # KEY COMMAND (plot_mtm.m)
  print MATLAB "plot_mtm(data,syn,dataf,synf,new,new_dt,dlnA,dt,dlnA_full,dt_full,dlnA_ave,dt_ave,new_cc,new_cc_dt,dlnA_cc,dt_cc,flims,$iall);\n";
  print MATLAB "orient tall; suptitle(title);\n";
  #print MATLAB "print ('-depsc2','$inputdir/${filebase}_measure.eps');\n";
  print MATLAB "print ('-depsc2','$epsfile');\n";
  print MATLAB "close;\n";
  #  print MATLAB "plot_mtm_quality('$inputdir');\n";

  print "\nplotted to: $epsfile \n";

}

print MATLAB "\n";
close(MATLAB);

File: utils/ADJOINT_TOMOGRAPHY_TOOLS/measure_adj/scripts_meas/plot_mtm.m
function plot_mtm(data,syn,dataf,synf,new,new_dt,dlnA,dt,dlnA_full,dt_full,dlnA_ave,dt_ave,...
		  new_cc,new_cc_dt,dlnA_cc,dt_cc,flims,iall)

fsize = 9;
lsize = 1.0;
msize = 12;

cgreen = [0.1210 0.6050 0];

fmax = max(flims);
f1 = -0.05*fmax;
f2 = 1.05*fmax;

% strings for cross-correlation measurements
stdT_cc = sprintf(' dT-cc (green) = %.3f +/- %.3f',dt_cc(1),dt_cc(2));
stdA_cc = sprintf(' dlnA-cc (green) = %.3f +/- %.3f',dlnA_cc(1),dlnA_cc(2));
stdT_mtave = sprintf(' dT-mt-ave (blue) = %.3f +/- %.3f',dt_ave(1),dt_ave(2));
stdA_mtave = sprintf(' dlnA-mt-ave (blue) = %.3f +/- %.3f',dlnA_ave(1),dlnA_ave(2));

subplot(3,1,1), hold on
%title('Am','fontsize',fsize)
plot(data(:,1),data(:,2),'-','linewidth',2*lsize,'Color','k')
plot(syn(:,1),syn(:,2),'-','linewidth',2*lsize,'Color','r')

if 0==1
  plot(new_dt(:,1),new_dt(:,2),'--','linewidth',lsize,'Color','cyan')
  plot(new(:,1),new(:,2),'--','linewidth',lsize,'Color','cyan')
  plot(new_cc_dt(:,1),new_cc_dt(:,2),'--','linewidth',lsize,'Color',cgreen)
  plot(new_cc(:,1),new_cc(:,2),'--','linewidth',lsize,'Color',cgreen)
  legend('Data','Syn','Syn-MT-1','Syn-MT-2','Syn-CC-1','Syn-CC-2','location','BestOutside')
else
  plot(new(:,1),new(:,2),'--','linewidth',lsize,'Color','cyan')
  plot(new_cc(:,1),new_cc(:,2),'--','linewidth',lsize,'Color',cgreen)
  legend('Data','Syn','Syn-MT','Syn-CC','location','BestOutside')
end
xlabel('Time [s]','fontsize',fsize)

% plot power of data and synthetics
subplot(3,3,4), hold on
plot(dataf(:,1),dataf(:,2),'-','linewidth',2*lsize,'Color','k')
plot(synf(:,1),synf(:,2),'-','linewidth',2*lsize,'Color','r')
xlim([f1 f2])
ax2 = axis; [xmat, ymat] = vertlines(flims,ax2(3),ax2(4)); plot(xmat(:,1:4),ymat(:,1:4),'k',xmat(:,5:6),ymat(:,5:6),'r');
ylabel('Power','fontsize',fsize)
xlabel('Frequency [Hz]','fontsize',fsize)
title('Power of data and synthetics','fontsize',fsize)

subplot(3,3,5), hold on
% plot multitaper measurements
if iall == 1
errorbar(dlnA(:,1),dlnA(:,2),dlnA(:,3),'.','markersize',msize,'linewidth',lsize)
ax1 = axis;
plot(dlnA_full(:,1),dlnA_full(:,2),'b','linewidth',lsize/2)
plot(dlnA_full(:,1),dlnA_full(:,2)+dlnA_full(:,3),'r--','linewidth',lsize/2)  % top error
plot(dlnA_full(:,1),dlnA_full(:,2)-dlnA_full(:,3),'r--','linewidth',lsize/2)  % bottom error
axis([f1 f2 ax1(3:4)])
end
plot_bars(f1,f2,dlnA_ave(1),dlnA_ave(2),[0 0 1],lsize)  % mt average measurement
plot_bars(f1,f2,dlnA_cc(1),dlnA_cc(2),cgreen,lsize)     % cc measurement
ax2 = axis; [xmat, ymat] = vertlines(flims,ax2(3),ax2(4)); plot(xmat(:,1:4),ymat(:,1:4),'k',xmat(:,5:6),ymat(:,5:6),'r');
ylabel('\Delta lnA','fontsize',fsize)
xlabel('Frequency [Hz]','fontsize',fsize)
  title({'Amplitude measurement',stdA_cc,stdA_mtave},'fontsize',fsize)

subplot(3,3,6), hold on
% plot multitaper measurements
if iall==1
errorbar(dt(:,1),dt(:,2),dt(:,3),'.','markersize',msize,'linewidth',lsize)
ax1 = axis;
plot(dt_full(:,1),dt_full(:,2),'b','linewidth',lsize/2)
plot(dt_full(:,1),dt_full(:,2)+dt_full(:,3),'r--','linewidth',lsize/2)  % top error
plot(dt_full(:,1),dt_full(:,2)-dt_full(:,3),'r--','linewidth',lsize/2)  % bottom error
axis([f1 f2 ax1(3:4)])
end
plot_bars(f1,f2,dt_ave(1),dt_ave(2),[0 0 1],lsize)  % mt average measurement
plot_bars(f1,f2,dt_cc(1),dt_cc(2),cgreen,lsize)     % cc measurement
ax2 = axis; [xmat, ymat] = vertlines(flims,ax2(3),ax2(4)); plot(xmat(:,1:4),ymat(:,1:4),'k',xmat(:,5:6),ymat(:,5:6),'r');
xlabel('Frequency [Hz]','fontsize',fsize)
ylabel('\Delta T [s]','fontsize',fsize)
title({'Traveltime measurement',stdT_cc,stdT_mtave},'fontsize',fsize)

